drivers/src/imsic/core.rs

```diff
@@ -550,37 +550,45 @@ impl Imsic {
         Ok(iter)
     }
 
-    /// Returns the IMSIC location specifier of the supervisor level interrupt file for the given
-    /// CPU.
-    pub fn supervisor_file_location(&self, cpu: CpuId) -> Result<ImsicLocation> {
+    /// Returns the IMSIC location specifier of the interrupt file on the specified physical CPU.
+    pub fn phys_file_location(&self, cpu: CpuId, file: ImsicFileId) -> Result<ImsicLocation> {
         let cpus = self.per_cpu.lock();
         let pcpu = cpus.get_cpu(cpu).ok_or(Error::InvalidCpu(cpu))?;
-        Ok(ImsicLocation::new(
-            pcpu.group,
-            pcpu.hart,
-            ImsicFileId::supervisor(),
-        ))
+        Ok(ImsicLocation::new(pcpu.group, pcpu.hart, file))
+    }
+
+    /// Returns the address of the interrupt file on the specified physical CPU.
+    pub fn phys_file_addr(&self, cpu: CpuId, file: ImsicFileId) -> Result<SupervisorPageAddr> {
+        self.phys_file_location(cpu, file).and_then(|loc| {
+            self.geometry
+                .location_to_addr(loc)
+                .ok_or(Error::InvalidGuestFile)
+        })
     }
 
     /// Returns the phandle of this IMSIC's node in the device-tree.
     pub fn phandle(&self) -> u32 {
         self.phandle
     }
 
-    /// Sends an IPI to the specified CPU.
-    pub fn send_ipi(&self, cpu: CpuId) -> Result<()> {
-        // If the CPU is valid, then its IMSIC location must be valid.
-        let addr = self
-            .geometry
-            .location_to_addr(self.supervisor_file_location(cpu)?)
-            .unwrap();
+    /// Sends an IPI with the raw `id` to the specified CPU and interrupt file.
+    pub fn send_ipi_raw(&self, cpu: CpuId, file: ImsicFileId, id: u32) -> Result<()> {
```

Comment:
Is this limited to IPI interrupts? Or can it inject any imsic interrupt id? Should a type other than u32 be used to indicate that?

Reply:
A CPU writing to the memory-mapped `seteipnum` is sort of by definition an IPI since it's software running on the CPU explicitly triggering an interrupt. I used `u32` here since MSI writes are always 4 bytes.

Reply:
sgtm